spec/models/relationship_spec.rb

```diff
@@ -23,4 +27,67 @@
       expect(@rel).to     be_filtered(["Host"], ["VmOrTemplate"])
     end
   end
+
+  describe ".filter_by_resource_type" do
+    let(:storages) { FactoryGirl.build_list(:relationship_storage_vmware, 1) }
+    let(:vms) { FactoryGirl.build_list(:relationship_vm_vmware, 1) }
+    let(:hosts) { FactoryGirl.build_list(:relationship_host_vmware, 1) }
+
+    it "includes" do
+      expect(filter(vms + hosts, :of_type => "Host")).to eq(hosts)
+      expect(filter(vms + hosts, :of_type => ["Host"])).to eq(hosts)
+      expect(filter(vms + hosts, :of_type => ["Host"], :except_type => [])).to eq(hosts)
+    end
+
+    it "includes multi" do
+      expect(filter(vms + hosts + storages, :of_type => %w(Host VmOrTemplate))).to match_array(vms + hosts)
+    end
+
+    it "includes everything" do
+      expect(filter(vms + hosts, :of_type => %w(Host VmOrTemplate))).to eq(vms + hosts)
+    end
+
+    it "includes nothing" do
+      expect(filter(vms, :of_type => ["Host"])).to be_empty
+    end
+
+    it "excludes" do
+      expect(filter(vms + hosts, :except_type => "Host")).to eq(vms)
+      expect(filter(vms + hosts, :except_type => ["Host"])).to eq(vms)
+      expect(filter(vms + hosts, :except_type => ["Host"], :of_type => [])).to eq(vms)
+    end
+
+    it "excludes multi" do
+      expect(filter(vms + hosts + storages, :except_type => %w(Host VmOrTemplate))).to eq(storages)
+    end
+
+    it "excludes everything" do
+      expect(filter(vms + hosts, :except_type => %w(Host VmOrTemplate))).to be_empty
+    end
+
+    it "excludes nothing" do
+      expect(filter(vms, :except_type => %w(Host))).to eq(vms)
+    end
+
+    it "includes and excludes" do
+      expect(filter(vms + hosts, :of_type => ["VmOrTemplate"], :except_type => ["Host"])).to eq(vms)
+    end
+
+    it "neither includes nor excludes" do
+      expect(filter(vms, {})).to eq(vms)
+    end
+
+    it "scopes" do
+      vms.map(&:save!)
+      hosts.map(&:save!)
+      storages.map(&:save!)
+      filtered_results = filter(Relationship.all, :of_type => %w(Host VmOrTemplate), :except_type => %w(Storage))
+      expect(filtered_results).not_to be_kind_of(Array)
+      expect(filtered_results).to match_array(vms + hosts)
+    end
+  end
+
+  def filter(relationships, options)
+    Relationship.filter_by_resource_type(relationships, options)
+  end
```

Comment:
I'm not a fan of moving the method-under-test into a helper method...yes, it's less typing, but the separation is confusing.